File: 3271. Hash Divided String.py
class Solution(object):
    def stringHash(self, s, k):
        i = 0                # Initialize index `i` to 0
        ans = ''             # Initialize an empty string `ans` to store the final result
        n = len(s)           # Get the length of the input string `s`
        
        # Loop through the string `s` until the end
        while i < n:
            sum = 0          # Initialize the sum for the current group of `k` characters
            temp = k         # Initialize a temporary variable `temp` to `k`
            
            # Sum the values of `k` characters in the current group
            while temp:
                sum += ord(s[i]) - ord('a')  # Convert the character to its corresponding 0-based index (e.g., 'a' -> 0)
                i += 1                        # Move to the next character
                temp -= 1                     # Decrease `temp` by 1 until `k` characters are processed
            
            # Calculate the index of the resulting character in the alphabet
            idx = (sum % 26) % 26
            
            # Convert the index back to a character and append it to the result string `ans`
            ans += chr(idx + ord('a'))
        
        # Return the final hashed string
        return ans

File: 3271. Hash Divided String.cpp
class Solution {
public:
    // Function to generate a hashed string from the input string 's' using integer 'k'
    string stringHash(string s, int k) 
    {
        string ans; // Initialize an empty string to store the resulting hash
        int i = 0, n = s.size(); // 'i' is the index iterator, 'n' is the length of the input string 's'
        
        // Iterate through the entire string
        while(i < n)
        {
            int sum = 0; // Variable to store the cumulative sum of ASCII values of 'k' characters
            int temp = k; // 'temp' is a temporary variable to keep track of the number of characters to sum

            // Loop to sum the ASCII values of the next 'k' characters
            while(temp--)
                sum += s[i++] - 'a'; // Convert character to an integer (0 for 'a', 1 for 'b', etc.) and add to sum
            
            // Calculate the index by taking modulo 26 twice, ensuring it wraps within the alphabet
            int idx = (sum % 26) % 26; 

            // Convert the index back to a character and append it to the result string 'ans'
            ans += idx + 'a';
        }

        // Return the final hashed string
        return ans;
    }
};